src/Microsoft.FileFormats/PDB/PDBFile.cs

```diff
@@ -8,23 +8,110 @@ namespace Microsoft.FileFormats.PDB
+        /// or MSFZ (compressed PDB). If this field is null, then this PDBFile is invalid.
+        /// </summary>
+        private readonly IMSFFile _msfFile;
+
         private readonly Lazy<Reader[]> _streams;
         private readonly Lazy<PDBNameStream> _nameStream;
         private readonly Lazy<DbiStream> _dbiStream;
 
         public PDBFile(IAddressSpace dataSource)
         {
-            _reader = new Reader(dataSource);
-            _header = new Lazy<PDBFileHeader>(() => _reader.Read<PDBFileHeader>(0));
+            Reader reader = new(dataSource);
             _streams = new Lazy<Reader[]>(ReadDirectory);
-            _nameStream = new Lazy<PDBNameStream>(() => new PDBNameStream(Streams[1]));
-            _dbiStream = new Lazy<DbiStream>(() => new DbiStream(Streams[3]));
+            _nameStream = new Lazy<PDBNameStream>(() => {
+                CheckValid();
+                return new PDBNameStream(_msfFile.GetStream(1));
+            });
+            _dbiStream = new Lazy<DbiStream>(() => {
+                CheckValid();
+                return new DbiStream(_msfFile.GetStream(3));
+            });
+
+            if (reader.Length > reader.SizeOf<PDBFileHeader>())
+            {
+                PDBFileHeader msfFileHeader = reader.Read<PDBFileHeader>(0);
+                if (msfFileHeader.IsMagicValid)
+                {
+                    MSFFile msfFile = MSFFile.OpenInternal(reader, msfFileHeader);
+                    if (msfFile != null)
+                    {
+                        _msfFile = msfFile;
+                        return;
+                    }
+                }
+            }
+
+            if (reader.Length > reader.SizeOf<MSFZFileHeader>())
+            {
+                MSFZFileHeader msfzFileHeader = reader.Read<MSFZFileHeader>(0);
+                if (msfzFileHeader.IsMagicValid)
+                {
+                    MSFZFile msfzFile = MSFZFile.Open(dataSource);
+                    if (msfzFile != null)
+                    {
+                        _msfFile = msfzFile;
+                        return;
+                    }
+                }
+            }
         }
 
-        public PDBFileHeader Header { get { return _header.Value; } }
+        /// <summary>
+        /// Opens a PDB file. If the operation is not successful, this method will throw an exception, rather
+        /// than return an invalid PDBFile object.
+        /// </summary>
+        public static PDBFile Open(IAddressSpace dataSource)
+        {
+            PDBFile pdbFile = new(dataSource);
+            if (pdbFile.IsValid())
+            {
+                return pdbFile;
+            }
+            else
+            {
+                throw new BadInputFormatException("The specified file is not a PDB (uses neither MSF nor MSFZ container format).");
+            }
+        }
+
+        [Obsolete("Switch to using NumStreams and GetStream. The 'Streams' collection is inefficient.")]
```

Comment:
@mikem8361 - can we do a quick search for code in our repos that consumes this API? If we aren't using it I'd propose we just delete it.

Reply:
We don't really know all the code internally and especially not externally that uses this code.  The old and new symbol uploaders don't because they the Microsoft.SymbolStore API which is implemented on top of this code.

Reply:
I think for now we leave it as Obsolete and then delete it in a future update.